From b80a6ef7dd0277d1654ec502883bfe59c1f4c264 Mon Sep 17 00:00:00 2001
From: Phil Steitz <psteitz@apache.org>
Date: Mon, 4 Feb 2008 03:58:08 +0000
Subject: [PATCH] Added exception handler to ensure that PooledConnections are
 not orphaned when an exception occurs in setUpDefaults or clearWarnings in
 IntanceKeyDataSource getConnection.

JIRA: DBCP-237
Reported and patched by Oliver Matz


git-svn-id: https://svn.apache.org/repos/asf/commons/proper/dbcp/trunk@618163 13f79535-47bb-0310-9956-ffa450edef68
---
 pom.xml                                       |  1 +
 project.xml                                   |  3 +-
 .../datasources/InstanceKeyDataSource.java    | 18 ++++-
 .../TestInstanceKeyDataSource.java            | 81 +++++++++++++++++++
 xdocs/changes.xml                             |  7 +-
 5 files changed, 104 insertions(+), 6 deletions(-)
 create mode 100644 src/test/org/apache/commons/dbcp/datasources/TestInstanceKeyDataSource.java

diff --git a/pom.xml b/pom.xml
index fc35539673..f33d6f2b43 100644
--- a/pom.xml
+++ b/pom.xml
@@ -255,6 +255,7 @@
                 <include>org/apache/commons/dbcp/datasources/TestKeyedCPDSConnectionFactory.java</include>
                 <include>org/apache/commons/dbcp/datasources/TestPerUserPoolDataSource.java</include>
                 <include>org/apache/commons/dbcp/datasources/TestSharedPoolDataSource.java</include>
+                <include>org/apache/commons/dbcp/datasources/TestInstanceKeyDataSource.java</include>
 
                 <include>org/apache/commons/dbcp/managed/TestBasicManagedDataSource.java</include>
                 <include>org/apache/commons/dbcp/managed/TestManagedDataSource.java</include>
diff --git a/project.xml b/project.xml
index 069fde2b5f..a3f079b607 100644
--- a/project.xml
+++ b/project.xml
@@ -366,7 +366,8 @@
         <include>org/apache/commons/dbcp/datasources/TestFactory.java</include>
         <include>org/apache/commons/dbcp/datasources/TestPerUserPoolDataSource.java</include>
         <include>org/apache/commons/dbcp/datasources/TestSharedPoolDataSource.java</include>
-
+        <include>org/apache/commons/dbcp/datasources/TestInstanceKeyDataSource.java</include>
+        
         <include>org/apache/commons/dbcp/managed/TestBasicManagedDataSource.java</include>
         <include>org/apache/commons/dbcp/managed/TestManagedDataSource.java</include>
         <include>org/apache/commons/dbcp/managed/TestManagedDataSourceInTx.java</include>
diff --git a/src/java/org/apache/commons/dbcp/datasources/InstanceKeyDataSource.java b/src/java/org/apache/commons/dbcp/datasources/InstanceKeyDataSource.java
index 6a330c230a..ad8bb245df 100644
--- a/src/java/org/apache/commons/dbcp/datasources/InstanceKeyDataSource.java
+++ b/src/java/org/apache/commons/dbcp/datasources/InstanceKeyDataSource.java
@@ -688,10 +688,20 @@ public Connection getConnection(String username, String password)
                                    + " to create the PooledConnection.");
         }
 
-        Connection con = info.getPooledConnection().getConnection();        
-        setupDefaults(con, username);
-        con.clearWarnings();
-        return con;
+        Connection con = info.getPooledConnection().getConnection();
+        try { 
+            setupDefaults(con, username);
+            con.clearWarnings();
+            return con;
+        } catch (SQLException ex) {  
+            try {
+                con.close();
+            } catch (Exception exc) { 
+                getLogWriter().println(
+                     "ignoring exception during close: " + exc);
+            }
+            throw ex;
+        }
     }
 
     protected abstract PooledConnectionAndInfo 
diff --git a/src/test/org/apache/commons/dbcp/datasources/TestInstanceKeyDataSource.java b/src/test/org/apache/commons/dbcp/datasources/TestInstanceKeyDataSource.java
new file mode 100644
index 0000000000..60384c6709
--- /dev/null
+++ b/src/test/org/apache/commons/dbcp/datasources/TestInstanceKeyDataSource.java
@@ -0,0 +1,81 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.commons.dbcp.datasources;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+
+import junit.framework.Test;
+import junit.framework.TestCase;
+import junit.framework.TestSuite;
+
+import org.apache.commons.dbcp.cpdsadapter.DriverAdapterCPDS;
+
+/**
+ * @version $Revision$ $Date$
+ */
+public class TestInstanceKeyDataSource extends TestCase {
+    public TestInstanceKeyDataSource(String testName) {
+        super(testName);
+    }
+
+    public static Test suite() {
+        return new TestSuite(TestInstanceKeyDataSource.class);
+    }
+
+
+    public void setUp() throws Exception {
+    }
+    
+    /**
+     * Verify that exception on setupDefaults does not leak PooledConnection
+     * 
+     * JIRA: DBCP-237
+     */
+    public void testExceptionOnSetupDefaults() throws Exception {
+        DriverAdapterCPDS pcds;
+        pcds = new DriverAdapterCPDS();
+        pcds.setDriver("org.apache.commons.dbcp.TesterDriver");
+        pcds.setUrl("jdbc:apache:commons:testdriver");
+        pcds.setUser("foo");
+        pcds.setPassword("bar");
+        pcds.setPoolPreparedStatements(false);
+        ThrowOnSetupDefaultsDataSource tds = new ThrowOnSetupDefaultsDataSource();
+        tds.setConnectionPoolDataSource(pcds);
+        int numConnections = tds.getNumActive();
+        try {
+            Connection conn = tds.getConnection("foo", "bar");
+            fail("Expecting SQLException");
+        } catch (SQLException ex) {
+           //Expected
+        }
+        assertEquals(numConnections,tds.getNumActive());     
+    }
+    
+    private static class ThrowOnSetupDefaultsDataSource
+    extends SharedPoolDataSource {
+        ThrowOnSetupDefaultsDataSource() {
+            super();
+        }
+        protected void setupDefaults(Connection con, String username)
+        throws  SQLException {
+            throw new SQLException("bang!");
+        }
+    }
+
+}
diff --git a/xdocs/changes.xml b/xdocs/changes.xml
index 7b52ccfc09..7d19ebce77 100644
--- a/xdocs/changes.xml
+++ b/xdocs/changes.xml
@@ -104,7 +104,12 @@ The <action> type attribute can be add,update,fix,remove.
       <action dev="psteitz" type="fix" issue="DBCP-245" due-to="Michael Drechsel">
         Fixed error in SharedPoolDataSource causing incorrect passwords to be 
         stored under certain conditions.
-      </action>   
+      </action> 
+      <action dev="psteitz" type="fix" issue="DBCP-237" due-to="Oliver Matz">
+        Added exception handler to ensure that PooledConnections are not
+        orphaned when an exception occurs in setUpDefaults or clearWarnings in 
+        IntanceKeyDataSource.getConnection.
+      </action> 
     </release>
     <release version="1.2.2" date="2007-04-04"
       description="This is a maintenance release containing bug fixes
